aikit/models/ordinal.py

```diff
@@ -0,0 +1,364 @@
+
+        if self._mono_target:
+            y_int = y_encoded[:,0]
+            assert len(self._target_encoder.categories_) == 1
+        else:
+            y_int = y_encoded # I keep the multi-dimension            
+            assert len(self._target_encoder.categories_) == y.shape[1]
+
+        self.regressor.fit(X, y_int)
+        
+        return self
+
+
+    def predict(self, X):
+        y_hat = self.regressor.predict(X)     # call regressor
+        y_int_hat = (y_hat + 0.5).astype(np.int32)  # conversion to closest int
+        
+        y_hat = self._target_encoder.inverse_transform(make2dimensions(y_int_hat))        
+        
+        if self._mono_target:
+            y_hat = y_hat[:, 0]
+
+        return y_hat.astype(self._target_dtype)
+
+
+    def predict_proba(self, X):
+        y_hat = self.regressor.predict(X)     # call regressor
+        if self._mono_target:
+            y_hat_2d = y_hat[:, np.newaxis]
+            assert y_hat.ndim == 1
+        else:
+            y_hat_2d = y_hat
+            assert y_hat.ndim == 2
+            
+
+        pivot_integers = [np.arange(len(category))[np.newaxis, :] for category in self._target_encoder.categories_]
+        
+        distances_to_pivot = [np.abs(y_hat_2d[:, j:(j+1)] - pivot_integers[j]) for j in range(y_hat_2d.shape[1])] # Rmk [:, j:(j+1)]  so that I keep the dimension
+        
+        probas = [self.distance_to_proba(distance_to_pivot) for distance_to_pivot in distances_to_pivot]
+
+        if self._mono_target:
+            return probas[0]
+    
+        else:
+            return probas
+
+
+    def distance_to_proba(self, d):
+        """ convert a distance to a probability """
+        e = np.exp(-d/self.kernel_windows) # TODO : find a good heuristic for that kernel_windows
+        
+        return e / e.sum(axis=1, keepdims=True)
+
+
+class OrdinalClassifier(_BaseOrdinalClassifier):
+    """ This class transform a classifier to make it more suited to ordinal classification.
+    It does so by changing the Target using the OrdinalOneHotEncoder transformer
+    
+    Concretely if we have 4 ordered classes 'y=A', 'y=B', 'y=C', 'y=D' with ('A' < 'B' < 'C' < 'D') 
+    It creates 3 targets :
+        'y > A' , 'y>B' and 'y>C'
+        
+    The classifier is then fitted on those target.
+    
+    At the end to make a prediction we call the underlying classifier and recreates the proba
+    
+    """
+    def __init__(self, classifier, classes="auto"):
+        self.classifier = classifier
+        self.classes=classes
+ 
+       
+    def fit(self, X , y):
+        
+        if not is_classifier(self.classifier):
+            raise TypeError("classifier should be a sklearn classifier")
+
+        y_int = self._prepare_target(y, klass = _OrdinalOneHotEncoder, conversion_type=DataTypes.DataFrame)
+        y_int = convert_generic(y_int, output_type=DataTypes.NumpyArray)
+        
+        self.classifier.fit(X, y_int)
```

Comment:
here to be sure:
we use only one classifier and trained it on a "multiclass" classification problem ?
is it different of cloning the classifer and train each clone separately ?

Reply:
yes : one classifier trained on **all** the target (4 ordered class ==> 3 target).

So you need a model that works with multi target. And then 2 choices :
 1. your model deal with all the target together (like DecisionTree/Forest in sklearn)
2. your model doesn't and in that case you need to encapsulate your model with for example this object 
https://scikit-learn.org/stable/modules/generated/sklearn.multioutput.MultiOutputClassifier.html
(which I don't do by myself here)

Reply:
ok understood so if you don't encapsulate, can we make a check that the classifier manages multi output ?

Reply:
I don't know if there is a clean way to know that here ... in the auto-ml it will be part of the information we add in the register but here I don't think we can.
(I'll check again)